<a href="https://colab.research.google.com/github/kjp07/data-python-learning/blob/main/Describing_data.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Describing data with summary statistics
---

This worksheet is tied to the Pandas Getting Started Tutorials, picking out particular tutorials to link them into a theme here.

We will focus on describing data.  This is the least risky in terms of bias and inaccurate conclusions as it should focus just on what data is presented to us.

Each exercise will ask you to work through on tutorial on the Getting Started page, to try the code from the tutorial here and to try a second, similar action.

---

The practice data from the tutorials comes from a dataset on Titanic passengers.


### Exercise 1 - open the Titanic dataset
---

The Titanic dataset is stored at this URL:
https://raw.githubusercontent.com/pandas-dev/pandas/master/doc/data/titanic.csv

Read the dataset into a pandas dataframe that you will call **titanic**.

**Test output**:  
The shape of the dataframe will be (891, 12)

In [1]:
import pandas as pd

url =  "https://raw.githubusercontent.com/pandas-dev/pandas/master/doc/data/titanic.csv"
df = pd.read_csv(url)
df.shape

(891, 12)

### Exercise 2 - get summary information about the dataframe
---

Read through the tutorials:  
[What kind of data does pandas handle?](https://pandas.pydata.org/docs/getting_started/intro_tutorials/01_table_oriented.html#)  
[How do I read and write tabular data?](https://pandas.pydata.org/docs/getting_started/intro_tutorials/02_read_write.html)

Use panda functions to display the following:
1.  A technical summary of the data (info())
2.  A description of the numerical data (describe())
3.  Display the Series 'Age'

**Test output**:   
1.  The info should show that there are only 204 values in the Cabin series, out of 891 records.  
2.  The description should show 7 columns and a mean age of 29.699118
3.  The Age series should have values of type float64 and Length 891

In [2]:
# technical summary
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [3]:
# Descriptive stats
df.describe()

PassengerId    Survived      Pclass  ...       SibSp       Parch        Fare
count   891.000000  891.000000  891.000000  ...  891.000000  891.000000  891.000000
mean    446.000000    0.383838    2.308642  ...    0.523008    0.381594   32.204208
std     257.353842    0.486592    0.836071  ...    1.102743    0.806057   49.693429
min       1.000000    0.000000    1.000000  ...    0.000000    0.000000    0.000000
25%     223.500000    0.000000    2.000000  ...    0.000000    0.000000    7.910400
50%     446.000000    0.000000    3.000000  ...    0.000000    0.000000   14.454200
75%     668.500000    1.000000    3.000000  ...    1.000000    0.000000   31.000000
max     891.000000    1.000000    3.000000  ...    8.000000    6.000000  512.329200

[8 rows x 7 columns]

In [4]:
# Display "Age"
df["Age"]

0      22.0
1      38.0
2      26.0
3      35.0
4      35.0
       ... 
886    27.0
887    19.0
888     NaN
889    26.0
890    32.0
Name: Age, Length: 891, dtype: float64

### Exercise 3 - aggregating statistics
---

Read through the tutorial:  
[How to calculate summary statistics?](https://pandas.pydata.org/docs/getting_started/intro_tutorials/06_calculate_statistics.html#)  

Use panda functions to display the following summary statistics from the titanic dataset:  

1.  The average (mean) age of passengers  
2.  The median age and fare  
3.  The mean fare
4.  The modal fare and gender

**Test output**:   
29.699118, Age 28.0000 Fare 14.4542, 32.2042079685746, Fare 8.05 Sex male 


In [5]:
# Mean age of passengers
mean_age = round(df["Age"].mean(), 0)
print("The mean age of the passengers is", mean_age, "(rounded to 0 decimal places).")

The mean age of the passengers is 30.0 (rounded to 0 decimal places).


In [6]:
# Median age and fare of passengers
df[["Age", "Fare"]].median()

Age     28.0000
Fare    14.4542
dtype: float64

In [7]:
# Mean fare
df["Fare"].mean()

32.2042079685746

### Exercise 4 - displaying other statistics
---

Take a look at the list of methods available for giving summary statistics [here](https://pandas.pydata.org/docs/user_guide/basics.html#basics-stats) 

Use panda functions, and your existing knowledge, to display the following summary statistics from the titanic dataset:

1.  The total number of passengers on the titanic
2.  The age of the youngest passenger
3.  The most expensive ticket price
4.  The range of ticket prices
5.  The number of passenges with cabins
6.  The code for the port where the highest number of passengers embarked
7.  The most populous gender
8.  The standard deviation for age and fare

**Tests**:  
891, 0.42, 512.3292, 512.3292, 204, S, male, Age 14.526497 Fare 49.693429

In [8]:
# Total number of passengers
df["PassengerId"].count()

891

In [9]:
# Age of youngest passenger
df["Age"].min()

0.42

In [10]:
# Most expensive ticket price
df["Fare"].max()

512.3292

In [11]:
# Range of ticket prices
df["Fare"].max() - df["Fare"].min()

512.3292

In [12]:
# Number of passengers with cabins
df[["PassengerId", "Cabin"]].count()

PassengerId    891
Cabin          204
dtype: int64

In [13]:
# Code for the port where the highest number of passengers embarked
df.groupby("Embarked")["PassengerId"].count()

Embarked
C    168
Q     77
S    644
Name: PassengerId, dtype: int64

In [14]:
# Most populous gender
df.groupby("Sex")["PassengerId"].count()

Sex
female    314
male      577
Name: PassengerId, dtype: int64

In [15]:
# Standard deviation for age and fare
df[["Age", "Fare"]].std()

Age     14.526497
Fare    49.693429
dtype: float64

### Exercise 5 - aggregating statistics grouped by category
---

Refer again to the tutorial  
[How to calculate summary statistics?](https://pandas.pydata.org/docs/getting_started/intro_tutorials/06_calculate_statistics.html#)   
looking particularly at the section on Aggregating statistics grouped by category.

1.  What is the mean age for male versus female Titanic passengers?
2.  What is the mean ticket fare price for each of the sex and cabin class combinations?
3.  What is the mean ticket fare price for passengers who embarked at each port?
4.  Which passenger class had the highest number of survivors (for now, just show the statistics - it may not be meaningful yet)?

**Test output**:  
1.  female 27.915709 male 30.726645
2.  
```
female  1         106.125798
            2          21.970121
            3          16.118810
male    1          67.226127
            2          19.741782
            3          12.661633
```
3.  
```
C    59.954144
Q    13.276030
S    27.079812
```
4. 

```
Survived  Pclass

0         1          80
          2          97
          3         372
1         1         136
          2          87
          3         119
```






In [16]:
# Mean age for male versus female Titanic passengers
df.groupby("Sex")["Age"].mean()

Sex
female    27.915709
male      30.726645
Name: Age, dtype: float64

In [17]:
# Mean ticket fare price for each of the sex and cabin class combinations
df.groupby(["Sex", "Pclass"])["Fare"].mean()

Sex     Pclass
female  1         106.125798
        2          21.970121
        3          16.118810
male    1          67.226127
        2          19.741782
        3          12.661633
Name: Fare, dtype: float64

In [18]:
# Mean ticket fare price for passengers who embarked at each port
df.groupby("Embarked")["Fare"].mean()

Embarked
C    59.954144
Q    13.276030
S    27.079812
Name: Fare, dtype: float64

In [19]:
# Passenger class with the highest number of survivors (for now, just show the statistics - it may not be meaningful yet)
df.groupby(["Survived", "Pclass"])["PassengerId"].count()

Survived  Pclass
0         1          80
          2          97
          3         372
1         1         136
          2          87
          3         119
Name: PassengerId, dtype: int64

### Exercise 6 - an aggregation of different statistics
---

Use the function titanic.agg() as shown in the tutorial  
[How to calculate summary statistics?](https://pandas.pydata.org/docs/getting_started/intro_tutorials/06_calculate_statistics.html#)  

1.  Display

```
     {
         "Age": ["min", "max", "median", "skew"],
         "Fare": ["min", "max", "median", "mean"]
     }
```
2.  Display:  
min, max and mean for Age  
min, max and standard deviation for Fare  
count for Cabin

**Test output**:   
1.  	
```
                  Age	      Fare  
max	   80.000000	512.329200  
mean	  NaN	      32.204208  
median	28.000000	14.454200  
min	   0.420000	 0.000000  
skew	  0.389108	 NaN
```

2.   
```
	        Age	    Fare	   Cabin
count  NaN	    NaN	    204.0
max	80.000000  512.329200 NaN
mean   29.699118  NaN        NaN
min	0.420000   0.000000   NaN
std	NaN        49.693429  NaN
```




In [20]:
# Display summary stats for Age and Fare
df.agg({
    "Age": ["min", "max", "median", "skew"],
    "Fare": ["min", "max", "median", "mean"]
    })

Age        Fare
max     80.000000  512.329200
mean          NaN   32.204208
median  28.000000   14.454200
min      0.420000    0.000000
skew     0.389108         NaN

In [21]:
# Display:
# min, max and mean for Age
# min, max and standard deviation for Fare
# count for Cabin

df.agg({
    "Age": ["min", "max", "mean"],
    "Fare": ["min", "max", "std"],
    "Cabin": ["count"]
})

Age        Fare  Cabin
count        NaN         NaN  204.0
max    80.000000  512.329200    NaN
mean   29.699118         NaN    NaN
min     0.420000    0.000000    NaN
std          NaN   49.693429    NaN

### Exercise 7 - count by category
---

Read the section Count number of records by category in the tutorial  
[How to calculate summary statistics?](https://pandas.pydata.org/docs/getting_started/intro_tutorials/06_calculate_statistics.html#)

1. Display the number of passengers of each gender who had a ticket
2. Display the number of passengers who embarked at each port and had a ticket
3. Calculate the percentage of PassengerIds who survived the sinking of the Titanic (*Hint:  try getting the PassengerIds with a count for survived or not.  Store this value in a new variable, which will contain a list/array.  The second item in this list will be the number who survived.  You can use this number and the count of PassengerIds to calculate the percentage*)

**Test output**:  
1.  female 314, male 577
2.  C 168, Q 77, S 644
3.  38.38383838383838



In [22]:
# Number of passengers of each gender who had a ticket
df["Sex"].value_counts()

male      577
female    314
Name: Sex, dtype: int64

In [23]:
# Number of passengers who embarked at each port and had a ticket
df.groupby("Embarked")["Ticket"].count()

Embarked
C    168
Q     77
S    644
Name: Ticket, dtype: int64

In [24]:
# Calculate the percentage of PassengerIds who survived the sinking of the Titanic 
# (Hint: try getting the PassengerIds with a count for survived or not. Store this value in a new variable, which will contain a list/array. 
# The second item in this list will be the number who survived. You can use this number and the count of PassengerIds to calculate the percentage)

# Number of passengers
passenger_count = df["PassengerId"].count()

# Survival status
survived = df.groupby("Survived")["PassengerId"].count()
survived_count = survived[1]

# Calculate percentage
(survived_count / passenger_count) * 100

38.38383838383838

### Exercise 8 - summary happiness statistics
---

Open the data set here: https://github.com/futureCodersSE/working-with-data/blob/main/Happiness-Data/2019.xlsx?raw=true

It contains data on people's perception of happiness levels in a number of countries across the world.

1.  Display the number of records in the set  
2.  Display the description of the numerical data  
3.  Display the highest GDP and life expectancy  
4.  Display the mean, max and min for Freedom,  mean, max, min and skew for Generosity and mean, min, max and std for GDP  

**Test output**:  
1.  156
2.  Table showing count, mean, std, min, 25%, 50%, 75%, max for 8 columns
3.  GDP 0.905147, life expectancy 0.725244  
4.  


```
	   Freedom to make life choices	Generosity	GDP per capita
max	 0.631000	                   0.566000	  1.684000
mean	0.392571	                   0.184846	  0.905147
min	 0.000000	                   0.000000  	0.000000
skew	NaN	                        0.745942	  NaN
std	 NaN	                        NaN	       0.398389
```




In [25]:
url2 = "https://github.com/futureCodersSE/working-with-data/blob/main/Happiness-Data/2019.xlsx?raw=true"
happy = pd.read_excel(url2)
happy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 156 entries, 0 to 155
Data columns (total 9 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Overall rank                  156 non-null    int64  
 1   Country or region             156 non-null    object 
 2   Score                         156 non-null    float64
 3   GDP per capita                156 non-null    float64
 4   Social support                156 non-null    float64
 5   Healthy life expectancy       156 non-null    float64
 6   Freedom to make life choices  156 non-null    float64
 7   Generosity                    156 non-null    float64
 8   Perceptions of corruption     156 non-null    float64
dtypes: float64(7), int64(1), object(1)
memory usage: 11.1+ KB


In [26]:
happy.head(5)

Overall rank Country or region  ...  Generosity  Perceptions of corruption
0             1           Finland  ...       0.153                      0.393
1             2           Denmark  ...       0.252                      0.410
2             3            Norway  ...       0.271                      0.341
3             4           Iceland  ...       0.354                      0.118
4             5       Netherlands  ...       0.322                      0.298

[5 rows x 9 columns]

In [27]:
# Number of records
happy["Overall rank"].count()

156

In [28]:
# Description of the numerical data
happy.describe()

Overall rank       Score  ...  Generosity  Perceptions of corruption
count    156.000000  156.000000  ...  156.000000                 156.000000
mean      78.500000    5.407096  ...    0.184846                   0.110603
std       45.177428    1.113120  ...    0.095254                   0.094538
min        1.000000    2.853000  ...    0.000000                   0.000000
25%       39.750000    4.544500  ...    0.108750                   0.047000
50%       78.500000    5.379500  ...    0.177500                   0.085500
75%      117.250000    6.184500  ...    0.248250                   0.141250
max      156.000000    7.769000  ...    0.566000                   0.453000

[8 rows x 8 columns]

In [29]:
# Highest GDP and life expectancy
print(happy["GDP per capita"].max())
print(happy["Healthy life expectancy"].max())

1.684
1.141


In [30]:
# mean, max and min for Freedom, mean, max, min and skew for Generosity and mean, min, max and std for GDP

happy.agg({
    "Freedom to make life choices": ["mean", "max", "min"],
    "Generosity": ["mean", "max", "min", "skew"],
    "GDP per capita": ["mean", "min", "max", "std"]
})

Freedom to make life choices  Generosity  GDP per capita
max                       0.631000    0.566000        1.684000
mean                      0.392571    0.184846        0.905147
min                       0.000000    0.000000        0.000000
skew                           NaN    0.745942             NaN
std                            NaN         NaN        0.398389

### Exercise 9 - migration data
---

Open the dataset at this url: https://github.com/futureCodersSE/working-with-data/blob/main/Data%20sets/public_use-talent-migration.xlsx?raw=true  Open the sheet named *Country Migration*

1.  Describe the dataset  
2.  Show summary information
3.  Display the mean net per 10K migration in each of the years 2015 to 2019
4.  Display the mean, max and min migration for the year 2019 for each of the regions (*base_country_wb_region*)
5.  Display the median net migration for the years 2015 and 2019 for the base countries by income level
6.  Display the number of target countries in each income level
7.  Display the mean net migration for all five years, for each income level

**Test output**:  
1  count, mean, std, min, 25%, 50%, 75%, max for 9 columns

2  shows 16 columns with non-null count of 4148 in each column

3  
```
net_per_10K_2015    0.461757
net_per_10K_2016    0.150248
net_per_10K_2017   -0.080272
net_per_10K_2018   -0.040591
net_per_10K_2019   -0.022743
dtype: float64
```

4  
```
	net_per_10K_2019
mean	max	min
base_country_wb_region			
East Asia & Pacific	0.198827	21.57	-9.88
Europe & Central Asia	0.208974	87.71	-21.34
Latin America & Caribbean	-0.904602	21.15	-31.75
Middle East & North Africa	-0.107655	55.60	-50.33
North America	0.239246	23.20	-0.29
South Asia	-0.514577	13.72	-24.89
Sub-Saharan Africa	-0.279729	37.11	-21.54
```

5  
```
	net_per_10K_2015	net_per_10K_2019
base_country_wb_income		
High Income	0.02	0.04
Low Income	0.42	-0.05
Lower Middle Income	-0.02	-0.07
Upper Middle Income	-0.03	-0.08
```

6  
```
base_country_wb_income
High Income            2415
Low Income              185
Lower Middle Income     653
Upper Middle Income     895
Name: target_country_name, dtype: int64 
```

7  
```
net_per_10K_2015	net_per_10K_2016	net_per_10K_2017	net_per_10K_2018	net_per_10K_2019
base_country_wb_income					
High Income	0.505482	0.391379	0.314178	0.379201	0.401470
Low Income	1.876432	0.798270	-0.684865	-0.677784	-0.681459
Lower Middle Income	0.591654	-0.029893	-0.519433	-0.527136	-0.476616
Upper Middle Income	-0.043419	-0.502916	-0.699240	-0.686626	-0.700101
```






In [38]:
url3 = "https://github.com/futureCodersSE/working-with-data/blob/main/Data%20sets/public_use-talent-migration.xlsx?raw=true"
cm = pd.read_excel(url3, sheet_name="Country Migration")

# Show summary information
cm.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4148 entries, 0 to 4147
Data columns (total 17 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   base_country_code         4148 non-null   object 
 1   base_country_name         4148 non-null   object 
 2   base_lat                  4148 non-null   float64
 3   base_long                 4148 non-null   float64
 4   base_country_wb_income    4148 non-null   object 
 5   base_country_wb_region    4148 non-null   object 
 6   target_country_code       4148 non-null   object 
 7   target_country_name       4148 non-null   object 
 8   target_lat                4148 non-null   float64
 9   target_long               4148 non-null   float64
 10  target_country_wb_income  4148 non-null   object 
 11  target_country_wb_region  4148 non-null   object 
 12  net_per_10K_2015          4148 non-null   float64
 13  net_per_10K_2016          4148 non-null   float64
 14  net_per_

In [39]:
# Describe the dataset
cm.describe()

base_lat    base_long  ...  net_per_10K_2018  net_per_10K_2019
count  4148.000000  4148.000000  ...       4148.000000       4148.000000
mean     28.418022    21.698305  ...         -0.040591         -0.022743
std      25.086012    61.937381  ...          3.593876          3.633247
min     -40.900557  -106.346771  ...        -56.220000        -50.330000
25%      14.058324    -3.435973  ...         -0.210000         -0.210000
50%      35.861660    19.145136  ...          0.000000          0.000000
75%      47.516231    53.688046  ...          0.170000          0.180000
max      64.963051   179.414413  ...         91.410000         87.710000

[8 rows x 9 columns]

In [42]:
# Display the mean net per 10K migration in each of the years 2015 to 2019
cm[["net_per_10K_2015", "net_per_10K_2016", "net_per_10K_2017", "net_per_10K_2018", "net_per_10K_2019"]].mean()

net_per_10K_2015    0.461757
net_per_10K_2016    0.150248
net_per_10K_2017   -0.080272
net_per_10K_2018   -0.040591
net_per_10K_2019   -0.022743
dtype: float64

In [45]:
# Display the mean, max and min migration for the year 2019 for each of the regions (base_country_wb_region)
cm.groupby("base_country_wb_region").agg({"net_per_10K_2019": ["mean", "min", "max"]})

net_per_10K_2019              
                                       mean    min    max
base_country_wb_region                                   
East Asia & Pacific                0.198827  -9.88  21.57
Europe & Central Asia              0.208974 -21.34  87.71
Latin America & Caribbean         -0.904602 -31.75  21.15
Middle East & North Africa        -0.107655 -50.33  55.60
North America                      0.239246  -0.29  23.20
South Asia                        -0.514577 -24.89  13.72
Sub-Saharan Africa                -0.279729 -21.54  37.11

In [46]:
# Display the median net migration for the years 2015 and 2019 for the base countries by income level
cm.groupby("base_country_wb_income").agg({
    "net_per_10K_2015": ["median"],
    "net_per_10K_2019": ["median"]})

net_per_10K_2015 net_per_10K_2019
                                 median           median
base_country_wb_income                                  
High Income                        0.02             0.04
Low Income                         0.42            -0.05
Lower Middle Income               -0.02            -0.07
Upper Middle Income               -0.03            -0.08

In [48]:
# Display the number of target countries in each income level
cm.groupby("target_country_wb_income")["target_country_name"].count()

target_country_wb_income
High Income            2415
Low Income              185
Lower Middle Income     653
Upper Middle Income     895
Name: target_country_name, dtype: int64

In [49]:
# Display the mean net migration for all five years, for each income level
cm.groupby("base_country_wb_income")[["net_per_10K_2015", "net_per_10K_2016", "net_per_10K_2017", "net_per_10K_2018", "net_per_10K_2019"]].mean()

net_per_10K_2015  ...  net_per_10K_2019
base_country_wb_income                    ...                  
High Income                     0.505482  ...          0.401470
Low Income                      1.876432  ...         -0.681459
Lower Middle Income             0.591654  ...         -0.476616
Upper Middle Income            -0.043419  ...         -0.700101

[4 rows x 5 columns]

### Exercise 10 - calculating range over a grouped series

Open the dataset at this url: https://github.com/futureCodersSE/working-with-data/blob/main/Data%20sets/public_use-talent-migration.xlsx?raw=true Open the sheet named *Skill Migration*

1.  Display the max for each skill group category of net migration for the year 2017
2.  Assign the max for each skill group category of net migration for the year 2017 to a variable called **max_skill_migration** and print `max_skill_migration`
3.  Create a second variable called **min_skill_migration** and assign to it the min for each skill group category of net migration for the year 2017, print `min_skill_migration`

4.  You now have two series `max_skill_migration` and `min_skill_migration` each of which is a numpy array.  You can perfom calculations on these two series in the same way as you would individual data items.

So, you can calculate the range for each skill category by subtracting the `min_skill_migration` from `max_skill_migration` to get a new series **skill_migration_range**

skill_migration_range = max_skill_migration - min_skill_migration

Try it out.

5.  Now calculate the range for the year 2019
6.  Now calculate the range for countries grouped by base country income level for the year 2015

Test output:  
1 and 2  
```
skill_group_category
Business Skills                1048.20
Disruptive Tech Skills         1478.56
Soft Skills                    1572.35
Specialized Industry Skills    1906.14
Tech Skills                    1336.78
Name: net_per_10K_2017, dtype: float64
```

3    
```
skill_group_category
Business Skills               -3471.35
Disruptive Tech Skills        -2646.19
Soft Skills                   -2542.23
Specialized Industry Skills   -6604.67
Tech Skills                   -6060.98
Name: net_per_10K_2017, dtype: float64
```

4  
```
skill_group_category
Business Skills                4519.55
Disruptive Tech Skills         4124.75
Soft Skills                    4114.58
Specialized Industry Skills    8510.81
Tech Skills                    7397.76
Name: net_per_10K_2017, dtype: float64
```

5  
```
skill_group_category
Business Skills                4543.96
Disruptive Tech Skills         3651.81
Soft Skills                    5528.47
Specialized Industry Skills    4036.44
Tech Skills                    3424.45
Name: net_per_10K_2019, dtype: float64
```

6  
```
wb_income
High income            4246.50
Low income             4556.42
Lower middle income    2148.36
Upper middle income    4045.43
Name: net_per_10K_2015, dtype: float64
```





In [51]:
sm = pd.read_excel(url3, sheet_name="Skill Migration")
sm.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17617 entries, 0 to 17616
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   country_code          17617 non-null  object 
 1   country_name          17617 non-null  object 
 2   wb_income             17617 non-null  object 
 3   wb_region             17617 non-null  object 
 4   skill_group_id        17617 non-null  int64  
 5   skill_group_category  17617 non-null  object 
 6   skill_group_name      17617 non-null  object 
 7   net_per_10K_2015      17617 non-null  float64
 8   net_per_10K_2016      17617 non-null  float64
 9   net_per_10K_2017      17617 non-null  float64
 10  net_per_10K_2018      17617 non-null  float64
 11  net_per_10K_2019      17617 non-null  float64
dtypes: float64(5), int64(1), object(6)
memory usage: 1.6+ MB


In [52]:
# Max net migration for each skill group category in 2017
sm.groupby("skill_group_category")["net_per_10K_2017"].max()

skill_group_category
Business Skills                1048.20
Disruptive Tech Skills         1478.56
Soft Skills                    1572.35
Specialized Industry Skills    1906.14
Tech Skills                    1336.78
Name: net_per_10K_2017, dtype: float64

In [54]:
# Assign the max for each skill group category of net migration for the year 2017 to a variable called max_skill_migration and print max_skill_migration

print(max_skill_migration)

skill_group_category
Business Skills                1048.20
Disruptive Tech Skills         1478.56
Soft Skills                    1572.35
Specialized Industry Skills    1906.14
Tech Skills                    1336.78
Name: net_per_10K_2017, dtype: float64


In [55]:
# Create a second variable called min_skill_migration and assign to it the min for each skill group category of net migration for the year 2017, print min_skill_migration
min_skill_migration = sm.groupby("skill_group_category")["net_per_10K_2017"].min()
print(min_skill_migration)

skill_group_category
Business Skills               -3471.35
Disruptive Tech Skills        -2646.19
Soft Skills                   -2542.23
Specialized Industry Skills   -6604.67
Tech Skills                   -6060.98
Name: net_per_10K_2017, dtype: float64


In [57]:
# Skill migration range for 2017
skill_migration_range = max_skill_migration - min_skill_migration
print(skill_migration_range)

skill_group_category
Business Skills                4519.55
Disruptive Tech Skills         4124.75
Soft Skills                    4114.58
Specialized Industry Skills    8510.81
Tech Skills                    7397.76
Name: net_per_10K_2017, dtype: float64


In [58]:
# Now calculate the range for the year 2019
max_skill_migration_2019 = sm.groupby("skill_group_category")["net_per_10K_2019"].max()
min_skill_migration_2019 = sm.groupby("skill_group_category")["net_per_10K_2019"].min()
skill_migration_range_2019 = max_skill_migration_2019 - min_skill_migration_2019
print(skill_migration_range_2019)

skill_group_category
Business Skills                4543.96
Disruptive Tech Skills         3651.81
Soft Skills                    5528.47
Specialized Industry Skills    4036.44
Tech Skills                    3424.45
Name: net_per_10K_2019, dtype: float64


In [59]:
# Now calculate the range for countries grouped by base country income level for the year 2015
min_skill_migration_2015 = sm.groupby("wb_income")["net_per_10K_2015"].min()
max_skill_migration_2015 = sm.groupby("wb_income")["net_per_10K_2015"].max()
skill_migration_2015 = max_skill_migration_2015 - min_skill_migration_2015
print(skill_migration_2015)

wb_income
High income            4246.50
Low income             4556.42
Lower middle income    2148.36
Upper middle income    4045.43
Name: net_per_10K_2015, dtype: float64
